### PR TITLE
Allow ability to disable Redis capturing the args

### PR DESCRIPTION
**Pull request by callumj**
For security (& memory allocation reasons) we want to disable this as the keys could contain sensitive information. Allow for this to be disabled.

### CODE REVIEW

File: lib/ddtrace/contrib/redis/patcher.rb
Comment (ericmustin) [inline lib/ddtrace/contrib/redis/patcher.rb:67]: I think at the very least we want an `else` case where `span.resource = Datadog::Contrib::Redis::Ext::SPAN_COMMAND`. This fallback occurs at the agent-level in our ingest pipeline anyway afaik, but it's better to explicitly set it rather than having to rely on normalization code.

More broadly though, I think it's worth considering whether at least capturing the operation `get ? / set ? / etc` is preferable here to not capturing anything? This is what we do for `auth` commands currently, https://github.com/DataDog/dd-trace-rb/blob/750ebd9d820d433a80d011b2988bb26599c064a5/lib/ddtrace/contrib/redis/quantize.rb#L24.  If that can be accomplished in a performant way, would that still alleviate any security concerns? The `span.resource` provides a lot of value in the datadog UI, as we generate hits/latency/error statistics per (among other things) unique env/service/resource combination, and so if every resource is the same, the value of those redis service metrics are much lower.

Wdyt?

**Review by ericmustin — CHANGES_REQUESTED** [review]
@callumj Thanks for the contrib. A few nits and things I wanted to point out but generally this is something I'm on board with.

- I think our tracing ingest pipeline already has some hooks in place for the security aspect of this, https://docs.datadoghq.com/tracing/setup_overview/configure_data_security/?tab=redis#agent-trace-obfuscation where you can obfuscate `redis.raw_command` as well as the resource. Completely fine with addressing things at the tracing client level since it also has memory perf improvements, but more out of curiosity, have y'all tried that agent-based approach and found it lacking? Or just simply prefer to address things at the app level? If there were any issues with the agent approach, certainly that feedback is valuable, and I'd be happy to work with the folks maintaining the datadog-agent to try to improve things there.

- With the implementation as is, we're not setting _any_ resource, which is a bit problematic. I think at the datadog-agent level there's some normalization code that kicks in and assigns the `span.name` to be the resource if the resource is nil, but it's probably better to at least do that at the tracer level. If there's a happy path to setting just the operation `get / set / etc etc` i think that would be even better, but i'd like your opinion on this, i left a longer comment in the review.

Overall, great work, thanks for the contribution!

**Comment by callumj** [comment]
> I think our tracing ingest pipeline already has some hooks in place for the security aspect of this, https://docs.datadoghq.com/tracing/setup_overview/configure_data_security/?tab=redis#agent-trace-obfuscation where you can obfuscate redis.raw_command as well as the resource. Completely fine with addressing things at the tracing client level since it also has memory perf improvements, but more out of curiosity, have y'all tried that agent-based approach and found it lacking? Or just simply prefer to address things at the app level? If there were any issues with the agent approach, certainly that feedback is valuable, and I'd be happy to work with the folks maintaining the datadog-agent to try to improve things there.

Good call out - normally I would use the span filter as that is very powerful but in this area I figured it would be best to offer this as an option because you then avoid some allocations if you are using Redis really heavily and might be on an older interpreter. I would prefer controlling things at the app level because if we had a shared agent (like a sidecar) it would avoid config changes and instead just require simpler app level changes.

File: lib/ddtrace/contrib/redis/patcher.rb
Comment (callumj) [reply lib/ddtrace/contrib/redis/patcher.rb:67]: Good point, the first arg is pretty important. I'll play around with some approaches and add the best way in.

File: docs/GettingStarted.md
Comment (marcotc) [inline docs/GettingStarted.md:1430]: Nit documentation suggestion:
```suggestion
| `command_args` | Show the command arguments (e.g. `key` in `GET key`) as resource name and tag | true |
```

File: lib/ddtrace/contrib/redis/patcher.rb
Comment (marcotc) [reply lib/ddtrace/contrib/redis/patcher.rb:67]: I think the code path you are looking for is in `Datadog::Contrib::Redis::Quantize`.

You should be able to tweak it to be able to retrieve the command name, but not the arguments.

Just to reinforce what @ericmustin said, if the command is, for example, `ACL SETUSER user1 on allkeys +set`, we'd still like to capture `ACL SETUSER`.

**Review by marcotc — DISMISSED** [review]
Very nice work, @callumj! I added a few comments, but looks pretty good overall.

@ericmustin's feedback (which I see you have already read) has the main points we'd like to addressed to get it merged.

File: lib/ddtrace/contrib/redis/patcher.rb
Comment (callumj) [reply lib/ddtrace/contrib/redis/patcher.rb]: Thank you @marcotc & @ericmustin - just pushed some changes. Let me know what you think.

File: docs/GettingStarted.md
Comment (callumj) [reply docs/GettingStarted.md]: Renamed to `command_args` @marcotc

File: lib/ddtrace/contrib/redis/patcher.rb
Comment (marcotc) [inline lib/ddtrace/contrib/redis/patcher.rb:71]: Is there any reason `#get_verb` wouldn't work for pipelines?
```suggestion
                  commands = args[0].commands.map { |c| Datadog::Contrib::Redis::Quantize.get_verb(c) }
```

**Review by marcotc — COMMENTED** [review]
@callumj thank you for addressing our feedback, and quickly! The changes look good.

I left one comment about the `PIPELINE` implementation, please take a look and let me know if it makes sense.

File: lib/ddtrace/contrib/redis/patcher.rb
Comment (callumj) [reply lib/ddtrace/contrib/redis/patcher.rb:71]: Yeah good call.

File: lib/ddtrace/contrib/redis/patcher.rb
Comment (callumj) [inline lib/ddtrace/contrib/redis/patcher.rb:62]: @marcotc Assignment Branch Condition size has been violated because of this change: do you normally extract these out or ignore the cop?

File: lib/ddtrace/contrib/redis/ext.rb
Comment (marcotc) [inline lib/ddtrace/contrib/redis/ext.rb:14]: I think this can be removed now.
```suggestion
        PIPELINE_RESOURCE = 'PIPELINE'.freeze
```

File: lib/ddtrace/contrib/redis/patcher.rb
Comment (marcotc) [reply lib/ddtrace/contrib/redis/patcher.rb:62]: Your call, to be honest. I'm ok with either approach.

I noticed you extracted it, which actually looks much nicer!

Comment (callumj) [reply lib/ddtrace/contrib/redis/patcher.rb:62]: Sorry @marcotc the issue is in `patch_redis_client` (which does the `class_eval`).

Comment (marcotc) [reply lib/ddtrace/contrib/redis/patcher.rb:62]: Go ahead and add a "Assignment Branch Condition" exception here, no worries.

Comment (callumj) [reply lib/ddtrace/contrib/redis/patcher.rb:63]: Done, thank you @marcotc

**Review by ericmustin — APPROVED** [review]
lgtm, thank you for the contribution(s), @callumj . will wait for @marcotc to approve as well before merging, as he also had some feedback, although it looks like it's been addressed.

Thanks again!

**Review by marcotc — APPROVED** [review]
Thank you for your responsive updates, @callumj! 🚀 :shipit:

**Comment by callumj** [comment]
Thank you @marcotc, do you know when `0.44.0` will be released?

**Comment by marcotc** [comment]
@callumj we have one cross-team effort across all Datadog language tracers that we are almost ready to ship and we'd like to sync it with the next release. That work is we expect to be ready next week, so that's our timeline as of now.

**Comment by marcotc** [comment]
Thank you again for your work in this PR, @callumj. We've just released [v0.44.0](https://github.com/DataDog/dd-trace-rb/releases/tag/v0.44.0), which includes these changes. Let us know if you have any feedback with this new version.

**Pull request merged**
